File: 102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(!root) return {};
        std::vector<vector<int>> res;
        std::queue<TreeNode*> q;
        q.push(root);
        for(int N = q.size(); N != 0; N = q.size()){
            res.push_back(vector<int>());   //directly place the val in res can reduce the time of copy vector.
            for(int i = 0; i < N; i++){
                TreeNode* node = q.front();
                q.pop();
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
                res.back().push_back(node->val);
            }
        }
        return res;
    }
};
/*
1. create a deque
2. push the root to initialize the deuqe
loop
3. get the size n of deque.
4. iterate the node to push its child, and then pop the noden which is pushed to result array n times.
*/

File: 102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        Qeque<TreeNode> q = new LinkedList<>();
        q.addLast(root);
        for(int N = q.size(); N != 0; N = q.size()){
            res.add(new ArrayList<Integer>());   //directly place the val in res can reduce the time of copy vector.
            for(int i = 0; i < N; i++){
                TreeNode node = q.poll();
                if(node.left != null) q.offer(node.left);
                if(node.right != null) q.offer(node.right);
                res.get(res.size()-1).add(node.val);
            }
        }
        return res;
    }
}
